UNIXFSv1.md

```diff
@@ -0,0 +1,487 @@
+
+####### Duplicate names
+
+Duplicate names are not allowed, if two identical names are present in an directory, the decoder MUST error.
+
+##### `Symlink` type
+
+<!--TODO: check that this is true-->Symlinks MUST NOT have childs.
+
+Their Data.Data field is a path you may follow resolution at after appending it in front of the current path.
+
+###### Path resolution on symlinks
+
+Currently symlinks are not followable, that mean implementations needs to return symlinks objects and fail if a consumer tries to follow it through.
+
+This is a SHOULD level, you probably wont break much things if you start following them.
+
+##### `HAMTDirectory`
+
+<!-- TODO! -->
+
+`node.Data.hashType` indicates a multihash function to use to digest path components used for sharding.
+If this field is missing, murmur3-x64-64 (`0x22`) is implied.
+
+`node.Data.Data` is some bitfield, ones indicates weather or not the links are part of this HAMT or leaves of the HAMT.
+The usage of this field is unknown given you can deduce the same information from the links names.
+
+###### Path resolution on HAMTs
+
+Steps:
+1. Take the current path component then hash it using the multihash id provided in `Data.hashType`.
+2. Pop the lowest byte from the path component hash digest, hex encoded (using 0-F) this byte and find the link that starts with this hex encoded byte.
+3. if the link name is exactly two bytes (by that, only hex encoded element), follow the link and repeat step 2 with the child node. The child node MUST be a hamt directory else the directory is invalid. Else continue
+4. You successfully resolved a path component, everything past the hex encoded prefix is the name of that element (usefull when listing childs of this directory).
+
+
+##### `TSize` / `DagSize`
+
+This is an optional field for Links of `dag-pb` nodes, **it does not represent any meaningfull information of the underlying structure** and no known usage of it to this day (altho some implementation emit thoses).
+
+To compute the `dagsize` of a node (which would be stored in the parents) you sum the length of the dag-pb outside message binary length, plus the blocksizes of all child files.
+
+An example of where this could be usefull is as a hint to smart download clients, for example if you are downloading a file concurrently from two sources that have radically different speeds, it would probably be more efficient to download bigger links from the fastest source, and smaller ones from the slowest source.
+
+<!--TODO: check that this is true-->
+There is no failure mode known for this field, so your implementation should be able to decode nodes where this field is wrong (not the value you expect), partially or completely missing. This also allows smarter encoder to give a more accurate picture (for example don't count duplicate blocks, ...).
+
+##### Traversal order
+
+The traversal is the self node, then recursively traverse each link, from low index to high index.
```

Comment:
This area seems underspecified. What is it for/why is it relevant? Are these shoulds or musts? ....

My guess is that this is meant effectively a sibling spec to the UnixFS data structure itself that is for things like a UnixFS ADL used in a selector for GraphSync that insist on deterministic traversal orders. If so it should be marked as what it's for.

However, implementations may want to do traversals of UnixFS data in all sorts of ways:
- Downloading all entries of a HAMT in parallel and spitting out the results as they come in
- Directory entries sorted by lexicographical ordering instead of HAMT ordering
- Using breadth instead of depth first traversals if trying to accumulate information about the CIDs of blocks in the DAG quickly

Reply:
removed